lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp

```diff
@@ -577,6 +584,83 @@ CompRegCanonicalizer::matchAndRewrite(seq::CompRegOp op,
   return success();
 }
 
+LogicalResult
+MergeVectorizeOps::matchAndRewrite(VectorizeOp vecOp,
+                                   PatternRewriter &rewriter) const {
+  auto &currentBlock = vecOp.getBody().front();
+  IRMapping argMapping;
+  SmallVector<Value> newOperands;
+  SmallVector<VectorizeOp> vecOpsToRemove;
+  bool canBeMerged = false;
+  // Used to calculate the new positions of args after insertions and removals
+  unsigned paddedBy = 0;
+
+  for (unsigned argIdx = 0, numArgs = vecOp.getInputs().size();
+       argIdx < numArgs; ++argIdx) {
+    auto inputVec = vecOp.getInputs()[argIdx];
+    // Make sure that the input comes from a `VectorizeOp`
+    // Ensure that the input vector matches the output of the `otherVecOp`
+    // Make sure that the results of the otherVecOp have only one use
+    // Ignore ClockTypes for now
+    auto otherVecOp = inputVec[0].getDefiningOp<VectorizeOp>();
+    if (!otherVecOp || inputVec != otherVecOp.getResults() ||
+        !llvm::all_of(otherVecOp.getResults(),
+                      [](auto result) { return result.hasOneUse(); }) ||
+        llvm::any_of(otherVecOp.getBody().front().getArguments(), [](auto arg) {
+          return mlir::isa<circt::seq::ClockType>(arg.getType());
+        })) {
+      newOperands.insert(newOperands.end(), inputVec.begin(), inputVec.end());
+      continue;
+    }
+
+    canBeMerged = true;
+    newOperands.insert(newOperands.end(), otherVecOp.getOperands().begin(),
+                       otherVecOp.getOperands().end());
+
+    auto &otherBlock = otherVecOp.getBody().front();
+    for (auto otherArg : otherBlock.getArguments()) {
+      auto newArg = currentBlock.insertArgument(
+          argIdx + paddedBy, otherArg.getType(), otherArg.getLoc());
+      argMapping.map(otherArg, newArg);
+      ++paddedBy;
+    }
+
+    rewriter.setInsertionPointToStart(&currentBlock);
+    for (auto &op : otherBlock.without_terminator())
+      rewriter.clone(op, argMapping);
+
+    unsigned argNewPos = paddedBy + argIdx;
+    // Get the result of the return value and use it in all places the
+    // the `otherVecOp` results were used
+    auto retOp = cast<VectorizeReturnOp>(otherBlock.getTerminator());
+    rewriter.replaceAllUsesWith(currentBlock.getArgument(argNewPos),
+                                argMapping.lookupOrDefault(retOp.getValue()));
+    currentBlock.eraseArgument(argNewPos);
+    vecOpsToRemove.push_back(otherVecOp);
+    // We erased an arg so the padding decreased by 1
+    paddedBy--;
+  }
+
+  if (!canBeMerged)
+    return success();
```

Comment:
If the pattern cannot be applied, it should return `failure()`, but you need to make sure that no modification to the IR has already happened in that case.
Returning `success()` in such a case means that the operation is pushed onto the worklist again since the driver thinks the pattern was applied successfully and modified something. This leads to unnecessary inefficiencies.